### PR TITLE
Create qbittorrent-nox service

### PR DESCRIPTION
**Pull request by camilosampedro**
## Description of changes

The [`qbittorrent-nox`](https://search.nixos.org/packages?channel=23.11&show=qbittorrent-nox&from=0&size=50&sort=relevance&type=packages&query=qbittorrent) package already exists, which is a headless version of qbittorrent (Only accessible via web). It being a headless service makes it a good candidate for being configured via `services.*` options, so I opened this PR for adding some basic configurations for it to run as a Systemd service.

This is my first contribution to the project, so I will greatly appreciate any comments and suggestions!

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by camilosampedro** [comment]
I was able to test it on a clean install inside of a QEMU VM, but the whole VM seems to crash when I run the review command:

```
nix-shell -p nixpkgs-review --run "nixpkgs-review pr 279716"
```

This crashes the VM even when running without `sudo` permissions. I wonder if this is a bug and if there's anything to do about it.

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+  config = mkIf cfg.enable {
+
+    services.qbittorrent-nox.package = mkDefault (pkgs.qbittorrent-nox);
+
+    systemd.tmpfiles.rules = [
+      "d '${cfg.dataDir}' 0770 ${cfg.user} ${cfg.group}"
+      "d '${cfg.dataDir}/.config' 0770 ${cfg.user} ${cfg.group}"
+      "d '${cfg.dataDir}/.config/qBittorrent' 0770 ${cfg.user} ${cfg.group}"
+    ];
```
Comment (ambroisie) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:87]: Use `systemd.tmpfiles.settings`.

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    systemd.services.qbittorrent-nox = {
```
Comment (ambroisie) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:89]: Same note about renaming the service

```suggestion
    systemd.services.qbittorrent = {
```

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      after = [ "network.target" "local-fs.target" "network-online.target" "nss-lookup.target" ];
+      wantedBy = [ "multi-user.target" ];
+      path = [ cfg.package ];
```
Comment (ambroisie) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:92]: Not needed, better served by systemd's service hardening options.

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            --torrenting-port=${toString cfg.torrenting.port}
+        '';
+        Type = "exec";
+        User = cfg.user;
+        Group = cfg.group;
+        UMask = "0002";
+        PrivateTmp = "false";
+        TimeoutStopSec = 1800;
+      };
+      # preStart = preStart;
```
Comment (ambroisie) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:111]: Dead code.

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      (mkIf (cfg.torrenting.openFirewall) {
+        allowedTCPPorts = [ cfg.torrenting.port ];
+        allowedUDPPorts = [ cfg.torrenting.port ];
+      })
+      (mkIf (cfg.web.openFirewall) {
+        allowedTCPPorts = [ cfg.web.port ];
+      })
+    ];
+
+    environment.systemPackages = [ cfg.package ];
```
Comment (ambroisie) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:124]: Why?

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+    users.users = mkIf (cfg.user == "qbittorrent") {
+      qbittorrent = {
+        group = cfg.group;
+        uid = config.ids.uids.qbittorrent-nox;
```
Comment (ambroisie) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:129]: As explained, please refrain from adding more `uid`/`gid` values to NixOS.

```suggestion
```

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        home = cfg.dataDir;
+        description = "qbittorrent daemon user";
+      };
+    };
+
+    users.groups = mkIf (cfg.group == "qbittorrent") {
+      qbittorrent = {
+        gid = config.ids.gids.qbittorrent-nox;
```
Comment (ambroisie) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:137]: Same remark as for `uid`.

```suggestion
```

**Review by ambroisie — CHANGES_REQUESTED** [review]
What about a `downloadDir`? Is there any specific setup necessary there?

**Comment by ambroisie** [comment]
I'm also unsure whether we still need to use `lib.mdDoc` (maybe for backporting? I will have to check).

I would also argue that `with lib;` over the whole file is over-reaching. I'd be fine with using it in the `options = ...` block though.

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 8]
```diff
@@ -0,0 +1,141 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Comment (nu-nu-ko) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:3]: you should avoid this kind of use of with to address https://github.com/NixOS/nixpkgs/issues/208242 for this module.

use inherits where you find yourself excessively reusing `lib.foo`

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            default = false;
+            type = types.bool;
+            description = lib.mdDoc ''
+              Whether to open the firewall for the ports in
+              {option}`services.qbittorrent-nox.torrenting.port`.
+            '';
+          };
+        };
+
+        dataDir = mkOption {
```
Comment (nu-nu-ko) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:50]: is a dataDir true here for qbit, profiles seems to be the way it manages itself, a profilesdir commandline option is available.

File: nixos/modules/services/torrent/qbittorrent-nox.nix [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      path = [ cfg.package ];
+      unitConfig = {
+        Description = "qBittorrent-nox Daemon";
+        Documentation = "man:qbittorrent-nox(1)";
+      };
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/qbittorrent-nox \
+            --profile=${cfg.dataDir} \
+            --webui-port=${toString cfg.web.port} \
```
Comment (nu-nu-ko) [inline nixos/modules/services/torrent/qbittorrent-nox.nix:101]: neither port options should be done with the command line options, qbittorrent has  a configuration file we can define these in, we know where its state will be after all

**Comment by nu-nu-ko** [comment]
seems i should have looked harder before i started on this, see https://github.com/NixOS/nixpkgs/commit/69fa9574a7ebd9ace9a5ef917555b7f02b113907
differences between our work now are
1. qBittorrent.conf options thanks to `gendeepINI` ( thanks [eclairevoyant](https://github.com/eclairevoyant) )
2. I'm missing `openFirewall` ( i want to include in config values here )
3. I'm missing a type for the configs options.
4. I apply much more service hardening.

I'll draft my own PR if I can get `openFirewall` done in an acceptable way myself.
open to however you may want to work with this or not, would love to have the config options available with nix instead of expecting the user to write a file out themselves etc.

as reference for how my module can be used right now i have a working use case in my systems repo [here](https://github.com/nu-nu-ko/crystal/blob/main/mods/services/qbit.nix)

would also appreciate your insight as another current qbit user / someone who wants a service module for it.
as well as testing on `aarch-64`

**Comment by camilosampedro** [comment]
@nu-nu-ko Thank you for working on those changes, and sorry I've been absent these past days.

I'd be ok with you opening your PR, using that as the actual PR that introduces this service, and I can help testing it on my machine.

Let me know and I'll close this PR.

**Comment by nu-nu-ko** [comment]
I've created https://github.com/NixOS/nixpkgs/pull/287923 with the current issues listed, I have only somewhat addressed the read write directorys and open ports issues as defaults should be filled in properly.

please ensure that PR has the same goal and functionality as you do here before closing anything, thank you.

**Comment by camilosampedro** [comment]
@nu-nu-ko Your PR seems to be following my goal. I'll be doing some tests and commenting on that PR. I'll close this PR for now

**Pull request closed without merging**